pinot-core/src/main/java/org/apache/pinot/core/data/manager/offline/DimensionTableDataManager.java

```diff
@@ -0,0 +1,162 @@
+  private final Lock _lookupTableWriteLock = _rwl.writeLock();
+
+  // _lookupTable is a HashMap used for storing/serving records for a table keyed by table PK
+  @GuardedBy("_rwl")
+  private final Map<PrimaryKey, GenericRow> _lookupTable = new HashMap<>();
+  private Schema _tableSchema;
+  private List<String> _primaryKeyColumns;
+
+  @Override
+  protected void doInit() {
+    super.doInit();
+
+    _tableSchema = ZKMetadataProvider.getTableSchema(_propertyStore, _tableNameWithType);
+    Preconditions.checkState(_tableSchema != null, "Failed to find schema for table: %s", _tableNameWithType);
+    _primaryKeyColumns = _tableSchema.getPrimaryKeyColumns();
+    Preconditions.checkState(!CollectionUtils.isEmpty(_primaryKeyColumns),
+        "Primary key columns must be configured for dimension table: %s", _tableNameWithType);
+  }
+
+  @Override
+  public void addSegment(File indexDir, IndexLoadingConfig indexLoadingConfig)
+      throws Exception {
+    super.addSegment(indexDir, indexLoadingConfig);
+    try {
+      loadLookupTable();
+      _logger.info("Successfully loaded lookup table: {}", getTableName());
+    } catch (Exception e) {
+      throw new RuntimeException(String.format("Error loading lookup table: %s", getTableName()), e);
+    }
+  }
+
+  @Override
+  public void removeSegment(String segmentName) {
+    super.removeSegment(segmentName);
+    try {
+      loadLookupTable();
+      _logger.info("Successfully removed segment and reloaded lookup table: {}", getTableName());
+    } catch (Exception e) {
+      throw new RuntimeException(
+          String.format("Error reloading lookup table after segment remove for table: {}", getTableName()), e);
+    }
+  }
+
+  /**
+   * `loadLookupTable()` reads contents of the DimensionTable into _lookupTable HashMap for fast lookup.
+   */
+  private void loadLookupTable()
+      throws Exception {
+    _lookupTableWriteLock.lock();
+    try {
+      _lookupTable.clear();
+      List<SegmentDataManager> segmentManagers = acquireAllSegments();
+      if (segmentManagers.size() == 0) {
+        return;
+      }
+
+      for (SegmentDataManager segmentManager : segmentManagers) {
+        IndexSegment indexSegment = segmentManager.getSegment();
+        try (PinotSegmentRecordReader reader =
+            new PinotSegmentRecordReader(indexSegment.getSegmentMetadata().getIndexDir())) {
+          while (reader.hasNext()) {
+            GenericRow row = reader.next();
+            _lookupTable.put(row.getPrimaryKey(_primaryKeyColumns), row);
+          }
+        }
+        releaseSegment(segmentManager);
+      }
+    } finally {
+      _lookupTableWriteLock.unlock();
+    }
+  }
+
+  public GenericRow lookupRowByPrimaryKey(PrimaryKey pk) {
+    _lookupTableReadLock.lock();
+    try {
+      return _lookupTable.get(pk);
+    } finally {
+      _lookupTableReadLock.unlock();
+    }
+  }
+
+  public FieldSpec getColumnFieldSpec(String columnName) {
```

Comment:
Please annotate as visiblefortesting.

Reply:
Hi Subbu! This is actually one of the two public methods which, the proposed LookupUDF will use. So it is not visible for testing only. To see how the UDF will interact, please take a look at the POC implementation [here](https://github.com/cbalci/incubator-pinot/pull/1/files#diff-b3dc29283ab6ee4ae023e05d276284c754d5e0261df038e70a770f1899456d6bR115).

Reply:
OK